File: backend/app/schemas/faturamento.py
from datetime import datetime
from decimal import Decimal
from typing import Optional
from uuid import UUID
from pydantic import BaseModel, Field, field_validator


class FaturamentoCreate(BaseModel):
    ticket_id: UUID
    empresa_id: UUID
    valor: Decimal = Field(..., gt=0)
    descricao: Optional[str] = None
    mes_referencia: str = Field(..., pattern=r"^\d{4}-\d{2}$")
    faturado: bool = False
    numero_nota_fiscal: Optional[str] = None

    @field_validator("descricao")
    @classmethod
    def sanitize_descricao(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return None
        value = value.strip()
        return value or None


class FaturamentoUpdate(BaseModel):
    valor: Optional[Decimal] = Field(None, gt=0)
    descricao: Optional[str] = None
    mes_referencia: Optional[str] = Field(None, pattern=r"^\d{4}-\d{2}$")
    faturado: Optional[bool] = None
    numero_nota_fiscal: Optional[str] = None

    @field_validator("descricao")
    @classmethod
    def sanitize_descricao(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return None
        value = value.strip()
        return value or None


class FaturamentoResponse(BaseModel):
    id: UUID
    ticket_id: UUID
    empresa_id: UUID
    valor: Decimal
    descricao: Optional[str]
    mes_referencia: str
    data_faturamento: Optional[datetime]
    faturado: bool
    data_faturacao: Optional[datetime]
    numero_nota_fiscal: Optional[str]
    data_criacao: datetime
    data_atualizacao: datetime
    ativo: bool

    model_config = {"from_attributes": True}


class FaturamentoListItem(BaseModel):
    id: UUID
    ticket_id: UUID
    empresa_id: UUID
    ticket_numero: str
    ticket_titulo: str
    ticket_descricao: str
    categoria_nome: Optional[str] = None
    empresa_nome: Optional[str] = None
    data_ticket: datetime
    valor: Decimal
    descricao: Optional[str]
    mes_referencia: str
    faturado: bool
    data_faturacao: Optional[datetime]
    numero_nota_fiscal: Optional[str]


class FaturamentoResumo(BaseModel):
    mes_referencia: str
    total_registros: int
    subtotal_pendente: Decimal
    subtotal_faturado: Decimal
    total_geral: Decimal

File: frontend/src/api/faturamento.ts
import { api } from './client'

export interface FaturamentoItem {
  id: string
  ticket_id: string
  empresa_id: string
  ticket_numero: string
  ticket_titulo: string
  ticket_descricao: string
  categoria_nome?: string | null
  empresa_nome?: string | null
  data_ticket: string
  valor: number
  descricao?: string | null
  mes_referencia: string
  faturado: boolean
  data_faturacao?: string | null
  numero_nota_fiscal?: string | null
}

export interface FaturamentoResumo {
  mes_referencia: string
  total_registros: number
  subtotal_pendente: number
  subtotal_faturado: number
  total_geral: number
}

export interface FaturamentoCreateInput {
  ticket_id: string
  empresa_id: string
  valor: number
  descricao?: string
  mes_referencia: string
  faturado?: boolean
  numero_nota_fiscal?: string
}

export interface FaturamentoUpdateInput {
  valor?: number
  descricao?: string
  mes_referencia?: string
  faturado?: boolean
  numero_nota_fiscal?: string
}

export const faturamentoApi = {
  list: async (params?: {
    mes_referencia?: string
    empresa_id?: string
    faturado?: boolean
  }) => {
    const { data } = await api.get<FaturamentoItem[]>('/faturamento', { params })
    return data
  },

  resumo: async (params?: { mes_referencia?: string; empresa_id?: string }) => {
    const { data } = await api.get<FaturamentoResumo>('/faturamento/resumo', { params })
    return data
  },

  create: async (payload: FaturamentoCreateInput) => {
    const { data } = await api.post('/faturamento', payload)
    return data
  },

  update: async (id: string, payload: FaturamentoUpdateInput) => {
    const { data } = await api.put(`/faturamento/${id}`, payload)
    return data
  },

  updateStatus: async (id: string, faturado: boolean, numero_nota_fiscal?: string) => {
    const { data } = await api.patch(`/faturamento/${id}/status`, null, {
      params: { faturado, numero_nota_fiscal },
    })
    return data
  },

  remove: async (id: string) => {
    await api.delete(`/faturamento/${id}`)
  },

  exportCsv: async (params?: { mes_referencia?: string; empresa_id?: string; faturado?: boolean }) => {
    const response = await api.get('/faturamento/export/csv', {
      params,
      responseType: 'blob',
    })
    return response.data as Blob
  },

  exportJson: async (params?: { mes_referencia?: string; empresa_id?: string; faturado?: boolean }) => {
    const { data } = await api.get('/faturamento/export/json', { params })
    return data
  },
}
